From dd09cf9b2906e0aecfe4e3e29bda771e8bf55f95 Mon Sep 17 00:00:00 2001
From: Konstantin Porotchkin <kostap@marvell.com>
Date: Wed, 27 Jan 2021 17:11:13 +0200
Subject: [PATCH 1227/1921] Revert "Revert "arm64: dts: marvell: add SMMU
 support to CN913x family""

This reverts commit bf432e77fdaac92a44267a7fd980ac4169948aae.
Once the defconfig allows SMUU bypass by default, this patch could be
restored. Only DMA masters with configured IOMMU fields will go through
the ARM SMMU.

Signed-off-by: Konstantin Porotchkin <kostap@marvell.com>
Change-Id: Iaf38a5f99a18a5e6a4a3f9d4a801e7e2d3e85dcd
Reviewed-on: https://sj1git1.cavium.com/c/IP/SW/kernel/linux/+/44915
Tested-by: sa_ip-sw-jenkins <sa_ip-sw-jenkins@marvell.com>
Reviewed-by: Nadav Haklai <nadavh@marvell.com>
Reviewed-by: Stefan Chulski <stefanc@marvell.com>
[WK: The original patch got from Marvell sdk11.21.09]
Signed-off-by: Wenlin Kang <wenlin.kang@windriver.com>
---
 arch/arm64/boot/dts/marvell/cn9130-crb.dts |  5 +++++
 arch/arm64/boot/dts/marvell/cn9130-db.dts  |  5 +++++
 arch/arm64/boot/dts/marvell/cn9130.dtsi    | 21 +++++++++++++++++++++
 arch/arm64/boot/dts/marvell/cn9131-db.dts  |  2 ++
 4 files changed, 33 insertions(+)

diff --git a/arch/arm64/boot/dts/marvell/cn9130-crb.dts b/arch/arm64/boot/dts/marvell/cn9130-crb.dts
index 8ab3a770f4b8..22e04a2f3629 100644
--- a/arch/arm64/boot/dts/marvell/cn9130-crb.dts
+++ b/arch/arm64/boot/dts/marvell/cn9130-crb.dts
@@ -174,6 +174,11 @@
 		&cp0_comphy1 0
 		&cp0_comphy2 0
 		&cp0_comphy3 0>;
+	iommu-map =
+		<0x0   &smmu 0x480 0x20>,
+		<0x100 &smmu 0x4a0 0x20>,
+		<0x200 &smmu 0x4c0 0x20>;
+		iommu-map-mask = <0x031f>;
 };
 
 /* CON 28 */
diff --git a/arch/arm64/boot/dts/marvell/cn9130-db.dts b/arch/arm64/boot/dts/marvell/cn9130-db.dts
index 420b099fd468..640d80847f3b 100644
--- a/arch/arm64/boot/dts/marvell/cn9130-db.dts
+++ b/arch/arm64/boot/dts/marvell/cn9130-db.dts
@@ -277,6 +277,11 @@
 		&cp0_comphy1 0
 		&cp0_comphy2 0
 		&cp0_comphy3 0>;
+	iommu-map =
+		<0x0   &smmu 0x480 0x20>,
+		<0x100 &smmu 0x4a0 0x20>,
+		<0x200 &smmu 0x4c0 0x20>;
+		iommu-map-mask = <0x031f>;
 };
 
 &cp0_sata0 {
diff --git a/arch/arm64/boot/dts/marvell/cn9130.dtsi b/arch/arm64/boot/dts/marvell/cn9130.dtsi
index 0cd32d68a0e7..0835e3af2de3 100644
--- a/arch/arm64/boot/dts/marvell/cn9130.dtsi
+++ b/arch/arm64/boot/dts/marvell/cn9130.dtsi
@@ -37,3 +37,24 @@
 #undef CP11X_PCIE0_BASE
 #undef CP11X_PCIE1_BASE
 #undef CP11X_PCIE2_BASE
+
+&smmu {
+	status = "okay";
+};
+
+&cp0_sata0 {
+	iommus = <&smmu 0x444>;
+};
+
+&cp0_sdhci0 {
+	iommus = <&smmu 0x445>;
+};
+
+&cp0_usb3_0 {
+	iommus = <&smmu 0x440>;
+};
+
+&cp0_usb3_1 {
+	iommus = <&smmu 0x441>;
+};
+
diff --git a/arch/arm64/boot/dts/marvell/cn9131-db.dts b/arch/arm64/boot/dts/marvell/cn9131-db.dts
index 920b68cfb340..a67a3af02ce0 100644
--- a/arch/arm64/boot/dts/marvell/cn9131-db.dts
+++ b/arch/arm64/boot/dts/marvell/cn9131-db.dts
@@ -121,6 +121,7 @@
 
 &cp1_sata0 {
 	status = "okay";
+	iommus = <&smmu 0x454>;
 
 	/* CON32 */
 	sata-port@1 {
@@ -193,6 +194,7 @@
 /* CON58 */
 &cp1_usb3_1 {
 	status = "okay";
+	iommus = <&smmu 0x451>;
 	usb-phy = <&cp1_usb3_0_phy0>;
 	/* Generic PHY, providing serdes lanes */
 	phys = <&cp1_comphy3 1>;
-- 
2.31.1

